### PR TITLE
Mysql to spanner nb solution 

### LINKED ISSUE #205: [notebook] Migrating multiple tables from MySQL to Cloud Spanner
Create a jupyter notebook based solution to migrate MySQL to Cloud Spanner
Solution should allow for:
1. User to specify mysql connection.
2. Generate list of tables from metadata. Alternatively, user should be able to supply list of tables.
3. Identify current primary key column name, and partitioned read properties.
4. Should generate logic for partitioned read, such that each partition read is <2 GB in size.
5. Run JDBCToGCS (Python or Java template) for MySQL to GCS for export.
6. Run GCSToSpanner (Java) for import into Spanner.
7. Notebook should allow for both types of save modes i.e. appending data or overwrite
8. Notebook should allow table schema generation if table does not exists.

### PR DESCRIPTION
**Pull request by surjits254**
Jupyter notebook for migrating MySQL database to Spanner

1. User to specifies mysql connection.
2. Solution generates list of tables from metadata. Alternatively, user can provide list of tables.
3. Solution Identifies current primary key column name, User need to specify PK for tables which do not have primary key in mysql.
4.  uses JDBCToGCS Java for MySQL to GCS for export.
5. uses GCSToSpanner (Java) for import into Spanner.
6. allows append and overwrite modes
5. allows file types avro, parquet, orc 
6. GCSToSpanner creates table if does not exist in spanner since we are using file formats which have schema

created an issue for enhancement #270 
closes #205

### CODE REVIEW

File: python/notebooks/spanner/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+## Jupyter Notebook Solution for migrating MySQL database to lLoud Spanner using Dataproc Templates
```
Comment (NiloFreitas) [inline python/notebooks/spanner/README.md:1]: Fix typo: lLoud Spanner

File: python/notebooks/spanner/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+Template for reading tables from HIVE Database and writing to BigQuery Dataset.
```
Comment (NiloFreitas) [inline python/notebooks/spanner/README.md:3]: This description is about Hive, change to MySql/Spanner

File: python/notebooks/spanner/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+Refer [Setup Vertex AI - PySpark](./../README.md) to setup new Jupyter notebook in vertexAI. Once the setup is done navigate to 
+[dataproc-templates/python/notebooks/spanner](https://github.com/GoogleCloudPlatform/dataproc-templates/blob/mysql-to-spanner-nb/python/notebooks/spanner/) folder and open [mysql_to_spanner.ipynb](https://github.com/GoogleCloudPlatform/dataproc-templates/blob/mysql-to-spanner-nb/python/notebooks/spanner/mysql-to-spanner.ipynb) notebook.
```
Comment (NiloFreitas) [inline python/notebooks/spanner/README.md:6]: Change file name to the current naming convention: MySqlToSpanner_vertex_pipeline_pyspark.ipynb

File: python/notebooks/spanner/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#### GCS to SPANNER Parameters
+* `SPANNER_INSTANCE` : cloud spanner instance name
+* `SPANNER_DATABASE` : cloud spanner database name
+* `SPANNER_TABLE_PRIMARY_KEYS` : provide dictionary of format {"table_name":"primary_key"} for tables which do not have primary key in MYSQL
+
+### Required JAR files
+
+This notebook requires the MYSQL connector jar. Installation information is present in the notebook
+
+### Notebook contains step by step process for migrating MySQL database to Cloud Spanner
```
Comment (NiloFreitas) [inline python/notebooks/spanner/README.md:45]: Move this line as description in the first topic, and explain that the notebook does the migration in 2 steps: MySql to GCS and then GCS to Spanner

File: python/notebooks/spanner/mysql-to-spanner.ipynb [hunk 5]
```diff
@@ -0,0 +1,954 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bd8f6dd9-2e13-447c-b28d-10fa2321b759",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "PROJECT = <\"project-id\">\n",
```
Comment (NiloFreitas) [inline python/notebooks/spanner/mysql-to-spanner.ipynb:178]: Here, the characters " should be outside of the characters <>

File: python/notebooks/spanner/mysql-to-spanner.ipynb [hunk 6]
```diff
@@ -0,0 +1,954 @@
...
+    "JARS = [GCS_STAGING_LOCATION + \"/jars/mysql-connector-java-8.0.29.jar\",\"file:///usr/lib/spark/external/spark-avro.jar\"]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "051df2af-bd8b-47c7-8cb2-05404ca0d859",
+   "metadata": {},
+   "source": [
+    "### Step 3.2 MYSQL to GCS Parameters\n",
+    "#### MYSQL_HOST : MYSQL instance ip address\n",
```
Comment (NiloFreitas) [inline python/notebooks/spanner/mysql-to-spanner.ipynb:194]: Change the #### to paragraph-size font (without ####)

File: python/notebooks/spanner/mysql-to-spanner.ipynb [hunk 7]
```diff
@@ -0,0 +1,954 @@
...
+    "print(MYSQLTABLE_LIST)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1d9a62e8-7499-41c6-b32b-73b539b0c7c4",
+   "metadata": {},
+   "source": [
+    "## Step 5: Get Primary Keys for tables not present in SPANNER_TABLE_PRIMARY_KEYS\n",
+    "For tables which do not have primary key provided in dictonary SPANNER_TABLE_PRIMARY_KEYS this step fetches primary key from MYSQL_DATABASE"
```
Comment (NiloFreitas) [inline python/notebooks/spanner/mysql-to-spanner.ipynb:331]: Typo: dictonary

File: python/notebooks/spanner/mysql-to-spanner.ipynb [hunk 8]
```diff
@@ -0,0 +1,954 @@
...
+    "        SUBNETWORK_URIS: str = SUBNET,\n",
+    "        FILE_URIS: list = [GCS_STAGING_LOCATION + \"/\" + LOG4J_PROPERTIES]\n",
+    "    ):\n",
+    "        for table in EXECUTION_LIST:\n",
+    "            BATCH_ID = \"gcs2spanner-{}\".format(datetime.now().strftime(\"%s\"))\n",
+    "            gcs_to_spanner_jobs.append(BATCH_ID)\n",
+    "            TEMPLATE_SPARK_ARGS = [\n",
+    "            \"--template=GCSTOSPANNER\",\n",
+    "            \"--templateProperty\", \"project.id={}\".format(PROJECT),\n",
+    "            \"-templateProperty\",  \"gcs.spanner.input.format={}\".format(MYSQL_OUTPUT_GCS_FORMAT),\n",
```
Comment (NiloFreitas) [inline python/notebooks/spanner/mysql-to-spanner.ipynb:730]: Missing one more "-" character

**Review by NiloFreitas — CHANGES_REQUESTED** [review]
Nice work Surjit! I requested some changes

**Comment by surjits254** [comment]
Fixed as per the comments provided by @NiloFreitas

File: python/notebooks/spanner/README.md (hunk 1 above)
Comment (NiloFreitas) [reply python/notebooks/spanner/README.md]: Fix L

File: python/notebooks/spanner/mysql-to-spanner.ipynb [hunk 9]
```diff
@@ -0,0 +1,954 @@
...
+    "statusDF['gcs_to_spanner_status'] = gcs_to_spanner_status\n",
+    "statusDF"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0961f164-c7e4-4bb5-80f0-25fd1051147b",
+   "metadata": {},
+   "source": [
+    "## Step 12: Validate row counts of migrated tables from MySQL to SPANNER"
```
[the remark this answers is not in the document]
Comment (NiloFreitas) [reply python/notebooks/spanner/mysql-to-spanner.ipynb]: There are still other SPANNER words

**Review by NiloFreitas — APPROVED** [review]
LGTM, congrats!

**Pull request merged**